### ISSUE TITLE
New _batch_mahalanobis slower than in previous commit

### ISSUE THREAD

**Issue by danielcrane** [pytorch/pytorch#23928]
## 🐛 Bug

The [latest version](https://github.com/pytorch/pytorch/blob/master/torch/distributions/multivariate_normal.py#L23) of the `_batch_mahalanobis` function from `torch.distributions.multivariate_normal` is slower than that in a [previous commit](https://github.com/pytorch/pytorch/blob/bb7d1837bc164b06e1d0826a20a8f7e8338a44e3/torch/distributions/multivariate_normal.py#L59).

## To Reproduce

I made a comparison script to compare the two methods, which I've included below.

<details><summary>Comparison Script</summary>
<p>

```
import time
import torch
import numpy as np

def _batch_trtrs_lower(bb, bA):
    """
    Applies `torch.trtrs` for batches of matrices. `bb` and `bA` should have
    the same batch shape.
    """
    flat_b = bb.reshape((-1,) + bb.shape[-2:])
    flat_A = bA.reshape((-1,) + bA.shape[-2:])
    flat_X = torch.stack([torch.trtrs(b, A, upper=False)[0] for b, A in zip(flat_b, flat_A)])
    return flat_X.reshape(bb.shape)


def _batch_mahalanobis_old(bL, bx):
    r"""
    Computes the squared Mahalanobis distance :math:`\mathbf{x}^\top\mathbf{M}^{-1}\mathbf{x}`
    for a factored :math:`\mathbf{M} = \mathbf{L}\mathbf{L}^\top`.
    Accepts batches for both bL and bx. They are not necessarily assumed to have the same batch
    shape, but `bL` one should be able to broadcasted to `bx` one.
    """
    n = bx.size(-1)
    bL = bL.expand(bx.shape[bx.dim() - bL.dim() + 1:] + (n,))
    flat_L = bL.reshape(-1, n, n)  # shape = b x n x n
    flat_x = bx.reshape(-1, flat_L.size(0), n)  # shape = c x b x n
    flat_x_swap = flat_x.permute(1, 2, 0)  # shape = b x n x c
    M_swap = _batch_trtrs_lower(flat_x_swap, flat_L).pow(2).sum(-2)  # shape = b x c
    return M_swap.t().reshape(bx.shape[:-1])


def _batch_mv(bmat, bvec):
    r"""
    Performs a batched matrix-vector product, with compatible but different batch shapes.
    This function takes as input `bmat`, containing :math:`n \times n` matrices, and
    `bvec`, containing length :math:`n` vectors.
    Both `bmat` and `bvec` may have any number of leading dimensions, which correspond
    to a batch shape. They are not necessarily assumed to have the same batch shape,
    just ones which can be broadcasted.
    """
    return torch.matmul(bmat, bvec.unsqueeze(-1)).squeeze(-1)


def _batch_mahalanobis(bL, bx):
    r"""
    Computes the squared Mahalanobis distance :math:`\mathbf{x}^\top\mathbf{M}^{-1}\mathbf{x}`
    for a factored :math:`\mathbf{M} = \mathbf{L}\mathbf{L}^\top`.
    Accepts batches for both bL and bx. They are not necessarily assumed to have the same batch
    shape, but `bL` one should be able to broadcasted to `bx` one.
    """
    n = bx.size(-1)
    bx_batch_shape = bx.shape[:-1]

    # Assume that bL.shape = (i, 1, n, n), bx.shape = (..., i, j, n),
    # we are going to make bx have shape (..., 1, j,  i, 1, n) to apply batched tri.solve
    bx_batch_dims = len(bx_batch_shape)
    bL_batch_dims = bL.dim() - 2
    outer_batch_dims = bx_batch_dims - bL_batch_dims
    old_batch_dims = outer_batch_dims + bL_batch_dims
    new_batch_dims = outer_batch_dims + 2 * bL_batch_dims
    # Reshape bx with the shape (..., 1, i, j, 1, n)
    bx_new_shape = bx.shape[:outer_batch_dims]
    for (sL, sx) in zip(bL.shape[:-2], bx.shape[outer_batch_dims:-1]):
        bx_new_shape += (sx // sL, sL)
    bx_new_shape += (n,)
    bx = bx.reshape(bx_new_shape)
    # Permute bx to make it have shape (..., 1, j, i, 1, n)
    permute_dims = (list(range(outer_batch_dims)) +
                    list(range(outer_batch_dims, new_batch_dims, 2)) +
                    list(range(outer_batch_dims + 1, new_batch_dims, 2)) +
                    [new_batch_dims])
    bx = bx.permute(permute_dims)

    flat_L = bL.reshape(-1, n, n)  # shape = b x n x n
    flat_x = bx.reshape(-1, flat_L.size(0), n)  # shape = c x b x n
    flat_x_swap = flat_x.permute(1, 2, 0)  # shape = b x n x c
    M_swap = torch.triangular_solve(flat_x_swap, flat_L, upper=False)[0].pow(2).sum(-2)  # shape = b x c
    M = M_swap.t()  # shape = c x b

    # Now we revert the above reshape and permute operators.
    permuted_M = M.reshape(bx.shape[:-1])  # shape = (..., 1, j, i, 1)
    permute_inv_dims = list(range(outer_batch_dims))
    for i in range(bL_batch_dims):
        permute_inv_dims += [outer_batch_dims + i, old_batch_dims + i]
    reshaped_M = permuted_M.permute(permute_inv_dims)  # shape = (..., 1, i, j, 1)
    return reshaped_M.reshape(bx_batch_shape)


input_1 = torch.from_numpy(np.linalg.cholesky(np.diag(np.random.rand(2).astype(np.float32)))).cuda()
input_2 = torch.from_numpy(np.random.rand(13440, 2).astype(np.float32)).cuda()

runs = 1000

total_time_new, total_time_old = 0, 0
relative_error_cum = 0
for _ in range(runs):
    torch.cuda.synchronize()
    st = time.perf_counter()
    m_new = _batch_mahalanobis(input_1, input_2)
    torch.cuda.synchronize()
    total_time_new += time.perf_counter() - st

    st = time.perf_counter()
    m_old = _batch_mahalanobis_old(input_1, input_2)
    torch.cuda.synchronize()
    total_time_old += time.perf_counter() - st

    relative_error_cum = torch.norm(m_new-m_old) / torch.norm(m_new)
    
print(f'[New] Time Taken: {total_time_new:.6f}s')
print(f'[Old] Time Taken: {total_time_old:.6f}s')
print(f'Average Relative Error: {relative_error_cum/runs}')
```

</p>
</details>

The output I get when running this is:

```
[New] Time Taken: 0.677909s
[Old] Time Taken: 0.159045s
Average Relative Error: 7.209917385342379e-11
```
Since the average relative difference between the methods over the runs is low, it doesn't seem like any additional calculations that would improve accuracy are the cause of the loss of speed.

## Environment

<details>
<summary> Environment Information </summary>
<p>

```
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A

OS: Pop!_OS 18.10
GCC version: (Ubuntu 8.3.0-6ubuntu1~18.10) 8.3.0
CMake version: version 3.12.1

Python version: 3.7
Is CUDA available: N/A
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 1070 With Max-Q Design
Nvidia driver version: 410.78
cuDNN version: /usr/lib/cuda-10.0/lib64/libcudnn.so.7.4.1

Versions of relevant libraries:
[pip3] numpy==1.17.0
[pip3] torch==1.1.0
[pip3] torchvision==0.3.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.1                      144  
[conda] mkl-service               1.1.2            py37he904b0f_5  
[conda] mkl_fft                   1.0.6            py37hd81dba3_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0
```

</p>
</details>

**Comment by vishwakftw** [comment]
cc: @fehiepsi. I believe you made the changes. By the way, batched triangular_solve handles broadcasting, so please feel free to use it.

**Comment by fehiepsi** [comment]
Hmm, the change should not have such high regression in GPU (in CPU I didn't recognize the regression), my bad. Thanks for a clear repro code! I will take a closer look soon.

@vishwakftw Does `triangular_solve(b, A)` handle the case `A.shape = (10, 10)`, `b.shape = (1000, 10, 1)` without broadcasting `A` to `(1000, 10, 10)`? If it could, then we can remove the permute logic in `_batch_mahalanobis`.

**Comment by vishwakftw** [comment]
@fehiepsi It internally performs the broadcasting. Is this undesirable?

**Comment by fehiepsi** [comment]
Yes, it is undesirable. For such shapes of A, and b, I would like to make `b` have shape `(10, 1000)` and apply triangular solve for A and new b (so no broadcasting is triggered).

Btw, the regression here is caused by a performance issue of triangular_solve in GPU. Consider
```
import torch
x = torch.eye(2).cuda()
y = torch.eye(2).cuda()
bx = x.reshape(1, 2, 2)
by = y.reshape(1, 2, 2)

%%timeit
torch.cuda.synchronize()
torch.triangular_solve(x, y)

%%timeit
torch.cuda.synchronize()
torch.triangular_solve(bx, by)
```
which will return
```
45.4 µs ± 318 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
for the first case and
```
243 µs ± 3.81 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
for the second case.

I guess we can fix it in `triangular_solve` implementation. Or I can add an if/else check to squeeze the input when `batch_shape=(1,)`. What do you think?

**Comment by vishwakftw** [comment]
Thanks for probing into this, @fehiepsi.

I think it’s best that this is fixed in the implementation of triangular_solve. I’ll send a patch shortly.

**Comment by danielcrane** [comment]
Thanks so much for reacting so quickly to this @vishwakftw @fehiepsi and others, really impressive!

**Comment by danielcrane** [comment]
@vishwakftw May I ask if you know whether or not this issue should be fixed in the latest nightly build? I've just installed and tested but am getting still getting similar timings to those in the repro script.

**Comment by fehiepsi** [comment]
@danielcrane I think you can use the stable version 1.2. I test your repro script there and find that the new version is a bit faster than the old version (make sure that you separate the calculation for total_time_old and total_time_new).

**Comment by danielcrane** [comment]
> @danielcrane I think you can use the stable version 1.2. I test your repro script there and find that the new version is a bit faster than the old version (make sure that you separate the calculation for total_time_old and total_time_new).

Oh, I didn't realise it slipped into 1.2.0, that's great! Thanks so much!

It's weird that when I tested using 1.2.0 the timing is good (~0.1s / 1000 runs) but when I just tried with the nightly version it was slower (similar speed to pre-patch). Hopefully nobody undid your hard work!

**Comment by Balandat** [comment]
Should this be closed out then?
